sql: allow for uncached resolution of type descriptors

rohany opened the pull request:
This PR adds the ability to create a type (but not really use it
anywhere). It adjusts the name resolution paths to be able to handle
the requested kind of descriptor, and disallows paths that want
table descriptors from accidentally accessing a type descriptor.
Upon type creation it performs similar logic to table creation
to write entries in the namespace and descriptor tables.

It additionally renames a variety of name resolution methods
to be more clear about what kind of object is being accessed.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48086)
<!-- Reviewable:end -->

rohany committed:
sql: allow for uncached resolution of type descriptors

This PR adds the ability to create a type (but not really use it
anywhere). It adjusts the name resolution paths to be able to handle
the requested kind of descriptor, and disallows paths that want
table descriptors from accidentally accessing a type descriptor.
Upon type creation it performs similar logic to table creation
to write entries in the namespace and descriptor tables.

It additionally renames a variety of name resolution methods
to be more clear about what kind of object is being accessed.

Release note: None

rohany commented:
I wanted to get some initial feedback on this PR before performing more renames and cleanup, and adding some more tests.

Some things I noticed while pulling together prototypes:
* Do we need to be distinguishing between table and type names? I feel like a single `ObjectName` struct for use after resolution to complement the `UnresolvedObjectName` makes sense. Additionally, there are places where I want to construct an ObjectName without knowing what kind of object I have/am looking for.
* I'm not sure of a good way to avoid a secondary lookup to determine the object type when colliding in a create type or create table operation.

knz reviewed (commented):
No comment on the names you've selected :+1:

but I won't be reviewing this for correctness.

ajwerner reviewed (commented):
Let's add a version gate to this PR for the creation code. 

I need to come back and give this a more thorough review. 

> Reviewed 5 of 5 files at r1, 32 of 32 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-:-M66wrkVAiiGRAADZ82Z:b-abxz7v)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, and @rohany)

___
*[pkg/sql/logical_schema_accessors.go, line 79 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M66xExK5GIV7MNHXj-i:-M66xExK5GIV7MNHXj-j:b-h6vizh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fd593cf6831dd583ed1e5d31164ca6edb6a6050/pkg/sql/logical_schema_accessors.go#L79)):*
> ```Go
> 	txn *kv.Txn,
> 	settings *cluster.Settings,
> 	db, schema, object string,
> ```

Why don't you like having a type for this triple? 
___
*[pkg/sql/sqlbase/structured.go, line 4013 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M66xzExBVilPrL1SRt-:-M66xzExBVilPrL1SRt0:b-scwmcf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fd593cf6831dd583ed1e5d31164ca6edb6a6050/pkg/sql/sqlbase/structured.go#L4013)):*
<details><summary><i alt="summary">Quoted 15 lines of code…</i></summary>

> ```
> 
> // TypeDesc implements the ObjectDescriptor interface.
> func (desc *MutableTableDescriptor) TypeDesc() *TypeDescriptor {
> 	return nil
> }
> 
> // TableDesc implements the ObjectDescriptor interface.
> func (desc *ImmutableTableDescriptor) TableDesc() *TableDescriptor {
> 	return &desc.TableDescriptor
> }
> 
> // TypeDesc implements the ObjectDescriptor interface.
> func (desc *ImmutableTableDescriptor) TypeDesc() *TypeDescriptor {
> 	return nil
> }
> ```
</details>


What's going on here?


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48086)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @lucy-zhang)

___
*[pkg/sql/logical_schema_accessors.go, line 79 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M66xExK5GIV7MNHXj-i:-M67l_8gBpZB7RhxGggD:b-23xyuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fd593cf6831dd583ed1e5d31164ca6edb6a6050/pkg/sql/logical_schema_accessors.go#L79)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Why don't you like having a type for this triple? 
</blockquote></details>

It was *TableName, but that doesn't make much sense for a generic object accessor. Based on the way some of the callers also decompose the name into these parts before accessing, so it seems reasonable.
___
*[pkg/sql/sqlbase/structured.go, line 4013 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M66xzExBVilPrL1SRt-:-M67m5Aw4uLVxjVo51SD:bdiapz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fd593cf6831dd583ed1e5d31164ca6edb6a6050/pkg/sql/sqlbase/structured.go#L4013)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 
> // TypeDesc implements the ObjectDescriptor interface.
> func (desc *MutableTableDescriptor) TypeDesc() *TypeDescriptor {
> 	return nil
> }
> 
> // TableDesc implements the ObjectDescriptor interface.
> func (desc *ImmutableTableDescriptor) TableDesc() *TableDescriptor {
> 	return &desc.TableDescriptor
> }
> 
> // TypeDesc implements the ObjectDescriptor interface.
> func (desc *ImmutableTableDescriptor) TypeDesc() *TypeDescriptor {
> 	return nil
> }
> ```

What's going on here?
</blockquote></details>

The schema accessors return an interface :
```
// ObjectDescriptor provides table information for results from a name lookup.
type ObjectDescriptor interface {
	tree.NameResolutionResult

	// TableDesc returns the underlying table descriptor, or nil if the
	// descriptor is not a table backed object.
	TableDesc() *TableDescriptor

	// TypeDesc returns the underlying type descriptor, or nil if the
	// descriptor is not a type backed object.
	TypeDesc() *TypeDescriptor
}
```
Before this patch it only had `TableDescriptor`. but I've extended it and it's implementers to also have the `TypeDesc` method. An alternative is to remove both and then typeswitch, but this makes it easier to operate on mutable and immutable table descs.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (changes requested):
> Reviewed 36 of 39 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48086)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, and @rohany)

___
*[pkg/sql/alter_table.go, line 134 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M6fCurC9oI_ETFZseC9:-M6fCurC9oI_ETFZseCA:b9npogp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6bb1dc9445580090e023c0aef3621ee2c0649eb/pkg/sql/alter_table.go#L134)):*
> ```
> 		return errors.AssertionFailedf("%q was not resolved as a table", n.n.Table)
> 	}
> ```
Consider also logging the type with `%T` and `tn`
___
*[pkg/sql/descriptor.go, line 245 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M6fDEpR06UYKOnGKy1N:-M6fDEpR06UYKOnGKy1O:byn5nol) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6bb1dc9445580090e023c0aef3621ee2c0649eb/pkg/sql/descriptor.go#L245)):*
<details><summary><i alt="summary">Quoted 6 lines of code…</i></summary>

> ```
> 
> // getDescriptorByID looks up the descriptor for `id`, validates it,
> // and unmarshals it into `descriptor`.
> //
> // In most cases you'll want to use wrappers: `getDatabaseDescByID` or
> // `getTableDescByID`.
> ```
</details>

This comment is stale now
___
*[pkg/sql/namespace_test.go, line 117 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M6fE25U4QAiqkFDULyo:-M6fE25VBD6BLLIq6_iv:b-nl7w24) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6bb1dc9445580090e023c0aef3621ee2c0649eb/pkg/sql/namespace_test.go#L117)):*
> ```
> 	// TODO (rohany): This is pretty hacky, but needs to be done in order
> 	//  for the exact errors to show up below.
> ```
Can't we use a regexp or something to make this less hacky?
___
*[pkg/sql/physical_schema_accessors.go, line 47 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M6fEhiQAqbjmGoWq7DG:-M6fEhiQAqbjmGoWq7DH:b-9aqgke) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6bb1dc9445580090e023c0aef3621ee2c0649eb/pkg/sql/physical_schema_accessors.go#L47)):*
> ```Go
> type UncachedPhysicalAccessor struct {
> 	// Used to avoid allocations.
> 	tn TableName
> ```

I'm sort of meh on this optimization. Should we even be passing pointers through all of this? I guess at some point it gets boxed in an `ObjectName`?
___
*[pkg/sql/physical_schema_accessors.go, line 215 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M6fEs_5EA2sRpKlHMGq:-M6fEs_5EA2sRpKlHMGr:b-y9hzwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6bb1dc9445580090e023c0aef3621ee2c0649eb/pkg/sql/physical_schema_accessors.go#L215)):*
> ```Go
> 			if flags.Required {
> 				a.tn = tree.MakeTableNameWithSchema(tree.Name(db), tree.Name(schema), tree.Name(object))
> 				return nil, sqlbase.NewUndefinedObjectError(&a.tn, flags.DesiredObjectKind)
> ```

Isn't this going to be a problem if this error gets stored anywhere and then another resolution happens?


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48086)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @lucy-zhang)

___
*[pkg/sql/descriptor.go, line 245 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M6fDEpR06UYKOnGKy1N:-M6fOtnY50nVOOTvF1Ru:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6bb1dc9445580090e023c0aef3621ee2c0649eb/pkg/sql/descriptor.go#L245)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 
> // getDescriptorByID looks up the descriptor for `id`, validates it,
> // and unmarshals it into `descriptor`.
> //
> // In most cases you'll want to use wrappers: `getDatabaseDescByID` or
> // `getTableDescByID`.
> ```
This comment is stale now
</blockquote></details>

Done.
___
*[pkg/sql/namespace_test.go, line 117 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M6fE25U4QAiqkFDULyo:-M6fOv7042hkJWS4XQg2:b-b18obh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6bb1dc9445580090e023c0aef3621ee2c0649eb/pkg/sql/namespace_test.go#L117)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	// TODO (rohany): This is pretty hacky, but needs to be done in order
> 	//  for the exact errors to show up below.
> ```
Can't we use a regexp or something to make this less hacky?
</blockquote></details>

The errors weren't regexable -- for example, some errors would say that the name collided with a database which was incorrect -- this is because the descriptor ID's the test was writing weren't in line with the real descriptor id counter.
___
*[pkg/sql/physical_schema_accessors.go, line 47 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M6fEhiQAqbjmGoWq7DG:-M6fPKdI3JqaYo2s6F6Z:b-ulhace) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6bb1dc9445580090e023c0aef3621ee2c0649eb/pkg/sql/physical_schema_accessors.go#L47)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'm sort of meh on this optimization. Should we even be passing pointers through all of this? I guess at some point it gets boxed in an `ObjectName`?
</blockquote></details>

I carried this over from what was originally done in the planner to avoid allocations for table name. Does taking the address of this field to pass as an object name cause another allocation? If so, then theres no point of this. Most of these are being used to just format error messages correctly.
___
*[pkg/sql/physical_schema_accessors.go, line 215 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M6fEs_5EA2sRpKlHMGq:-M6fQRh20m70f-8fk5E7:b1lm1op) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6bb1dc9445580090e023c0aef3621ee2c0649eb/pkg/sql/physical_schema_accessors.go#L215)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Isn't this going to be a problem if this error gets stored anywhere and then another resolution happens?
</blockquote></details>

What are you worried about here? The table name is used just for serialization (its not held on to by the error).


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-:-M6f_AMY666j6PFM4exA:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, and @rohany)

___
*[pkg/sql/alter_table.go, line 136 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M6f_51rBHXeiT_ayEW4:-M6f_51sBrckfBinWRkd:bws3aid) ([raw file](https://github.com/cockroachdb/cockroach/blob/d05b29f7f92d5b53d79f053e323f8bba39a4b993/pkg/sql/alter_table.go#L136)):*
> ```Go
> 	if !ok {
> 		return errors.AssertionFailedf(
> 			"%q was not resolved as a table. Instead it was %T", resolved, resolved)
> ```

nit: we don't generally put full sentences. 
___
*[pkg/sql/physical_schema_accessors.go, line 47 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M6fEhiQAqbjmGoWq7DG:-M6fZsG4En_X1VJ7_8Pv:b-cno991) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6bb1dc9445580090e023c0aef3621ee2c0649eb/pkg/sql/physical_schema_accessors.go#L47)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I carried this over from what was originally done in the planner to avoid allocations for table name. Does taking the address of this field to pass as an object name cause another allocation? If so, then theres no point of this. Most of these are being used to just format error messages correctly.
</blockquote></details>

No, it doesn't cause another allocation. It's just a space for bugs. 
___
*[pkg/sql/physical_schema_accessors.go, line 215 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M6fEs_5EA2sRpKlHMGq:-M6fZVUz1TO1Z94Of2Jq:b-33c3s9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6bb1dc9445580090e023c0aef3621ee2c0649eb/pkg/sql/physical_schema_accessors.go#L215)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

What are you worried about here? The table name is used just for serialization (its not held on to by the error).
</blockquote></details>

Fair enough, it's just scary when these cache values are passed as pointers outside of the scope of their use. It's easy to imagine a future change that decides it's a good idea to retain the pointer for a more structured error. 

Given it's error code, I could see just passing the value.


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48086)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, and @rohany)

___
*[pkg/sql/physical_schema_accessors.go, line 215 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M6fEs_5EA2sRpKlHMGq:-M6fqXTH3xerTpxg_Atv:b-ays2lq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6bb1dc9445580090e023c0aef3621ee2c0649eb/pkg/sql/physical_schema_accessors.go#L215)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Fair enough, it's just scary when these cache values are passed as pointers outside of the scope of their use. It's easy to imagine a future change that decides it's a good idea to retain the pointer for a more structured error. 

Given it's error code, I could see just passing the value.
</blockquote></details>

OK, that seems reasonable. I'll change these functions to take in strings instead of node formatters. But I'll do it as a separate PR, this one is too big already. It'll probably also let me remove these fields on the structs.


<!-- Sent from Reviewable.io -->

thoszhang reviewed (approved):
> Reviewed 5 of 5 files at r1, 2 of 32 files at r2, 33 of 39 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48086)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rohany)

___
*[pkg/sql/create_type.go, line 104 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M6ffhmG9xSvIHyxnw4q:-M6ffhmG9xSvIHyxnw4r:b-un7xqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d05b29f7f92d5b53d79f053e323f8bba39a4b993/pkg/sql/create_type.go#L104)):*
> ```Go
> 	if !p.ExecCfg().Settings.Version.IsActive(params.ctx, clusterversion.VersionEnums) {
> 		return pgerror.Newf(pgcode.FeatureNotSupported,
> 			"all nodes are not the correct version for ENUM type creation")
> ```

nit: `not all nodes are`
___
*[pkg/sql/physical_schema_accessors.go, line 340 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48086#-M6fnz1k3qx3UF45F9KY:-M6fnz1k3qx3UF45F9KZ:b84satj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d05b29f7f92d5b53d79f053e323f8bba39a4b993/pkg/sql/physical_schema_accessors.go#L340)):*
> ```Go
> 		return table, err
> 	default:
> 		return nil, errors.AssertionFailedf("unknown desired object kind")
> ```

Can we log what the flag value was?


<!-- Sent from Reviewable.io -->

rohany commented:
TFTRs!

bors r+

rohany commented:
bors r-

rohany commented:
bors r+
